components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/typing.scala

```diff
@@ -67,12 +67,20 @@ object typing {
 
   }
 
-  case class TypedTaggedValue(underlying: SingleTypingResult, tag: String) extends SingleTypingResult {
+  sealed trait TypedObjectWithData extends SingleTypingResult {
+    def underlying: SingleTypingResult
+    def data: AdditionalDataValue
 
     override def objType: TypedClass = underlying.objType
+  }
 
-    override def display: String = s"${underlying.display} @ $tag"
+  case class TypedTaggedValue(underlying: SingleTypingResult, tag: String) extends TypedObjectWithData {
+    override def data: AdditionalDataValue = tag
+    override def display: String = s"${underlying.display} @ ${data.display}"
+  }
 
+  case class TypedObjectWithValue(underlying: SingleTypingResult, data: AdditionalDataValue) extends TypedObjectWithData {
+    override def display: String = s"${underlying.display}"
```

Comment:
maybe we do want to display value here? Like "${underlying.display} of value ... } or sth? Or is it just to make tests pass ATM?

Reply:
IMO we could have `s"${underlying.display} @ ${data.display}"` on `TypedObjectWithData` level. It will looks pretty good for both tags and values

Reply:
Yes, we should display value here. This is temporary so that I don't have to correct all test cases, that depend on this method, yet. But `TypedObjectWithValue` cannot have display format like `s"${underlying.display} @ ${data.display}" ` because it will not look right for strings like `"", " "`. I was thinking about `s${underlying.display}{${data.display}}`.
I was also thinking about displaying types only when they are neccessary (eg. when expecting `Int{17}`, in union type `String{a} | String{b}`), but it would require changing code in a lot of places.